# Exercise 1 (Learning to Interrupt?)


In [2]:
# to interrupt, press 'i' twice on the keyboard
while True:
    print('*', end=' ')


* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * 

KeyboardInterrupt: 

*****  
# Exercise 2 (Resetting & Restarting the Kernel)

In [3]:
print = 'PRINT'

In [1]:
# To restart kernel, press '0' twice
print('Hello World')

Hello World


******  
# Exercise 3 (Timing Stuff)

In [23]:
squares = [i**2 for i in range(10)] 
print(squares)
%timeit for i in range(10): squares

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
356 ns ± 41.8 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


Analysis: 
- 356ns is the time taken to 1 loop to finish, with an uncertainty of 41.8ns per run.
- the 41.8ns uncertainty is the standard deviation of 7,000,000 loops (7 runs x 1,000,000 loops)

In [22]:
squares = list()
for i in range(10):
    squares.append(i**2)
print (squares)
%timeit for i in range(10): squares

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]
335 ns ± 35.2 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)


Analysis:
- 335ns is the time taken for 1 loop to finish, with an uncertainty of 35.2ns per run.
- the 35.2ns uncertainty is the standard deviation of 7,000,000 loops (7 runs x 1,000,000 loops)

Further thoughts: 
- Why is it faster for the 'longer' code to process and output a result compared to the 'shorter' code? (335 < 356)
- Likewise, why is the 'longer' code processing and outputting at a more consistent rate as compared to the 'shorter' code? (35.2 < 41.8)